File: Application/src/main/java/by/gastrofest/service/GastroSetService.java
package by.gastrofest.service;

import by.gastrofest.dbo.GastroSetDbo;
import by.gastrofest.repository.GastroSetRepository;
import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

import static by.gastrofest.constant.MainConstants.DATA_THUMB_ATTR;
import static by.gastrofest.constant.MainConstants.HREF_PROPERTY;
import static by.gastrofest.constant.MainConstants.LIST_TAG;
import static by.gastrofest.constant.MainConstants.MANI_PAGE_URL;
import static by.gastrofest.constant.MainConstants.SRC_PROPERTY;
import static by.gastrofest.constant.ParticipantConstants.IMAGES_CLASS;
import static by.gastrofest.constant.ParticipantConstants.IMAGE_CLASS;
import static by.gastrofest.constant.ParticipantConstants.INFO_SUMMARY_CLASS;
import static by.gastrofest.constant.ParticipantConstants.POSITIVE_POSSIBILITY;
import static by.gastrofest.constant.ParticipantConstants.RESERVED_CLASS;
import static by.gastrofest.constant.ParticipantConstants.SET_INFO_CLASS;
import static by.gastrofest.constant.ParticipantConstants.TO_TAKE_CLASS;
import static by.gastrofest.utils.HttpUtil.getEncodedString;

@Service
@RequiredArgsConstructor

public class GastroSetService {

    private final GastroSetRepository repository;

    @Transactional
    @SuppressWarnings("UnusedReturnValue")
    public GastroSetDbo save(final GastroSetDbo gastroSetDbo) {
        return repository.findByUrl(gastroSetDbo.getUrl()).orElseGet(() -> repository.save(gastroSetDbo));
    }

    public GastroSetDbo extractGastroSetInfoFromMainPage(final Element element) {

        final var imageElement = element.getElementsByClass(IMAGE_CLASS).get(0);
        final var imageLink = imageElement.absUrl(SRC_PROPERTY).split("\\?")[0];
        final var imageBase64 = getEncodedString(imageLink);
        final var url = MANI_PAGE_URL + Objects.requireNonNull(imageElement.parent()).attr(HREF_PROPERTY);
        return new GastroSetDbo(imageBase64, imageLink, url);
    }

    public void updateGastroSetFromGastroSetPage(final Document gastroSetDocument, final GastroSetDbo gastroSet) {
        gastroSet.setWeight(executeWeight(gastroSetDocument));
        gastroSet.setMealsDescriptions(executeMealsDescriptions(gastroSetDocument));
        gastroSet.setMealsImages(executeMealsImages(gastroSetDocument));
        gastroSet.setEatOutside(getPossibility(gastroSetDocument, TO_TAKE_CLASS));
        gastroSet.setBookingPossibility(getPossibility(gastroSetDocument, RESERVED_CLASS));

    }

    private static int executeWeight(final Document participantInfoDocument) {
        return Integer.parseInt(
                participantInfoDocument.getElementsByClass(INFO_SUMMARY_CLASS).get(0)
                        .getElementsByClass(SET_INFO_CLASS).get(0)
                        .childNodes().get(1)
                        .childNodes().get(0)
                        .toString()
                        .split(": ")[1]
                        .split(" ")[0]);
    }

    private List<String> executeMealsDescriptions(final Document participantInfoDocument) {
        return participantInfoDocument.getElementsByClass(INFO_SUMMARY_CLASS).get(0)
                .getElementsByTag(LIST_TAG)
                .stream()
                .map(Element::text)
                .map(String::trim)
                .collect(Collectors.toList());
    }

    private List<String> executeMealsImages(final Document participantInfoDocument) {
        return participantInfoDocument.getElementsByClass(IMAGES_CLASS).get(0)
                .childNodes()
                .stream()
                .map(node -> node.attr(DATA_THUMB_ATTR))
                .collect(Collectors.toList());
    }

    private static boolean getPossibility(final Document participantInfoDocument, final String className) {
        return POSITIVE_POSSIBILITY.equals(
                participantInfoDocument.getElementsByClass(className).get(0)
                        .getElementsByClass(SET_INFO_CLASS).get(0)
                        .childNodes().get(0)
                        .toString().trim());
    }

}

File: Parser/src/main/kotlin/by/gastrofest/parser/service/GastroSetService.kt
package by.gastrofest.parser.service

import by.gastrofest.parser.constant.DATA_THUMB_ATTR
import by.gastrofest.parser.constant.DELIVERY_CLASS
import by.gastrofest.parser.constant.HREF_PROPERTY
import by.gastrofest.parser.constant.IMAGES_CLASS
import by.gastrofest.parser.constant.IMAGE_CLASS
import by.gastrofest.parser.constant.INFO_SUMMARY_CLASS
import by.gastrofest.parser.constant.LIST_TAG
import by.gastrofest.parser.constant.MAIN_PAGE_URL
import by.gastrofest.parser.constant.POSITIVE_POSSIBILITY
import by.gastrofest.parser.constant.REPLACE_WORDS
import by.gastrofest.parser.constant.RESERVED_CLASS
import by.gastrofest.parser.constant.SET_INFO_CLASS
import by.gastrofest.parser.constant.SRC_PROPERTY
import by.gastrofest.parser.constant.THEME_CLASS
import by.gastrofest.parser.constant.TO_TAKE_CLASS
import by.gastrofest.parser.constant.WEIGHT_WORD
import by.gastrofest.parser.model.GastroSet
import org.jsoup.nodes.Document
import org.jsoup.nodes.Element
import org.jsoup.nodes.Node
import org.jsoup.nodes.TextNode
import java.util.*
import java.util.stream.Collectors


fun extractGastroSetInfoFromMainPage(element: Element): GastroSet {
    val imageElement = element.getElementsByClass(IMAGE_CLASS)[0]
    val imageLink = imageElement.absUrl(SRC_PROPERTY).split("\\?")[0]
    val url: String = MAIN_PAGE_URL + Objects.requireNonNull<Element?>(imageElement.parent()).attr(HREF_PROPERTY)
    return GastroSet(imageLink = imageLink, url = url)
}

fun updateGastroSetFromGastroSetPage(gastroSetDocument: Document, gastroSet: GastroSet) {
    gastroSet.weight = executeWeight(gastroSetDocument)
    gastroSet.theme = executeTheme(gastroSetDocument)
    gastroSet.mealsDescriptions = executeMealsDescriptions(gastroSetDocument)
    gastroSet.mealsImages = executeMealsImages(gastroSetDocument)
    gastroSet.eatOutside = getPossibility(gastroSetDocument, TO_TAKE_CLASS)
    gastroSet.booking = getPossibility(gastroSetDocument, RESERVED_CLASS)
    gastroSet.delivery = getPossibility(gastroSetDocument, DELIVERY_CLASS)
}

private fun executeWeight(participantInfoDocument: Document): Int? {
    return participantInfoDocument.getElementsByClass(INFO_SUMMARY_CLASS)[0]
        .getElementsByClass(SET_INFO_CLASS)[0]
        .childNodes()
        .stream().filter { node: Node -> node.toString().lowercase(Locale.getDefault()).contains(WEIGHT_WORD) }
        .findAny()
        .map<List<Node>> { obj: Node -> obj.childNodes() }
        .map { weight: List<Node> -> weight[0] }
        .map { obj: Node -> obj.toString() }
        .map { string: String -> string.split(": ")[1] }
        .map { string: String -> string.split(" ")[0] }
        .map { s: String -> s.toInt() }
        .orElse(null)
}

fun executeTheme(gastroSetDocument: Document): String? {
    if (gastroSetDocument.getElementsByClass(THEME_CLASS).isEmpty()) {
        return null
    }
    return (gastroSetDocument.getElementsByClass(THEME_CLASS)[0]
        .getElementsByClass(SET_INFO_CLASS)[0]
        .childNodes()[1]
        .childNodes()[0] as TextNode)
        .text()
        .replace(REPLACE_WORDS.joinToString("|").toRegex(), "")
        .trim()
}

private fun executeMealsDescriptions(participantInfoDocument: Document): List<String> {
    return participantInfoDocument.getElementsByClass(INFO_SUMMARY_CLASS)[0]
        .getElementsByTag(LIST_TAG)
        .stream()
        .map { obj: Element -> obj.text() }
        .map { obj: String -> obj.trim() }
        .collect(Collectors.toList())
}

private fun executeMealsImages(participantInfoDocument: Document): List<String> {
    return participantInfoDocument.getElementsByClass(IMAGES_CLASS)[0]
        .childNodes()
        .stream()
        .map { node: Node -> node.attr(DATA_THUMB_ATTR) }
        .collect(Collectors.toList())
}

private fun getPossibility(participantInfoDocument: Document, className: String): Boolean? {
    if (participantInfoDocument.getElementsByClass(className).isEmpty()) {
        return null
    }
    return POSITIVE_POSSIBILITY == participantInfoDocument.getElementsByClass(className)[0]
        .getElementsByClass(SET_INFO_CLASS)[0]
        .childNodes()[0]
        .toString()
        .trim()
}
